# project 1, classification digits : 손글씨 분류
## 1. 데이터 탐색 및 이해
### 1-1. 데이터 준비

In [75]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [76]:
digits = load_digits() # 데이터셋 할당

In [77]:
print(dir(digits)) # digits 내 변수 및 메소드 확인

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']


In [78]:
digits.keys() # 메서드 'keys'를 통한 내부 정보 확인

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [79]:
digits_data = digits.data # data 할당
print(digits_data.shape) # 배열 형상정보 출력

(1797, 64)


In [80]:
digits_data[0] # 0번째 데이터 64개 정보 출력

array([ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
       15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
       12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
        0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
       10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.])

In [81]:
digits_label = digits.target # 데이터별 lable 출력
print(digits_label)

[0 1 2 ... 8 9 8]


In [82]:
print(digits_label.shape)

(1797,)


In [83]:
digits.target_names # label명 확인

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [84]:
print(digits.DESCR) # 데이터셋 설명 확인

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

### 1-2. 학습 및 테스트 데이터 분리

In [85]:
from sklearn.model_selection import train_test_split
# 데이터셋 분리를 위한 train_test_split import

X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  1437 , X_test 개수:  360


In [86]:
X_train.shape, y_train.shape

((1437, 64), (1437,))

In [87]:
X_test.shape, y_test.shape

((360, 64), (360,))

## 2. 다양한 모델로 학습 및 평가하기
### 2-1. Decision Tree

In [88]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report # 결과값 레포트 함수 import

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred)) # 결과값 세부내용 확인

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



### 2-2. Random Forest

In [89]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report # 결과값 레포트 함수 import

random_forest = RandomForestClassifier(random_state=32) # 모델만 변경
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



### 2-3. Support Vector Machine (SVM)

In [90]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [91]:
from sklearn.metrics import classification_report # 수치 계산

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



### 2-4. Stochastic Gradient Descent Classifier (SGDClassifier)

In [92]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.82      0.95      0.88        42
           2       0.98      1.00      0.99        40
           3       0.97      0.91      0.94        34
           4       0.95      0.97      0.96        37
           5       0.93      1.00      0.97        28
           6       0.90      0.93      0.91        28
           7       0.89      0.97      0.93        33
           8       0.97      0.74      0.84        43
           9       0.97      0.88      0.92        32

    accuracy                           0.93       360
   macro avg       0.94      0.94      0.93       360
weighted avg       0.94      0.93      0.93       360



### 2-5. Logistic Regression

In [93]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


# project 1 summary
- project1은 classification project로, 총 1797개의 손글씨 이미지(8x8)를 64개 픽셀로 나누어 0~9까지 10개의 클래스로 분류시키는 모델을 구현했다.
- project에 사용된 모델은 5개이며 Decision Tree, Random Forest, Support Vector Machine(SVM), Stochastic Gradient Descent Classifier (SGDClassifier), Logistic Regressiond에 대해 평가 및 비교했다.
- classfication model 중 Random Forest의 accuracy가 0.96, confusion matrix의 macro avg가 0.96으로 정확도를 비롯해 각 lable별 precision, recall 값이 대체로 양호한 것을 확인할 수 있다. 반면에 Decision Tree는 accuracy가 비교적 낮은 것을 알 수 있다.
- project1은 이해를 목적으로 practice를 참고하여 진행하였으며, 다음 project2는 참고없이 진행하여 모델링 과정에 대해 숙달할 예정이다.